project/VersionUtil.scala

```diff
@@ -85,12 +86,27 @@ object VersionUtil {
     val date = executeTool("get-scala-commit-date")
     val sha = executeTool("get-scala-commit-sha").substring(0, 7)
 
+    val Patch = """\d+\.\d+\.(\d+)""".r
+    def cross = base match {
+      case Patch(p) if p.toInt > 0 => "bin"
+      case _ => "pre"
+    }
+
     val (canonicalV, mavenSuffix, osgiV, release) = suffix match {
-      case "SNAPSHOT"     => (s"$base-$date-$sha",   s"-SNAPSHOT",      s"$base.v$date-$sha",         false)
-      case "SHA-SNAPSHOT" => (s"$base-$date-$sha",   s"-$sha-SNAPSHOT", s"$base.v$date-$sha",         false)
-      case "SHA-NIGHTLY"  => (s"$base-$sha-nightly", s"-$sha-nightly",  s"$base.v$date-NIGHTLY-$sha", true)
-      case ""             => (s"$base",              "",                s"$base.v$date-VFINAL-$sha",  true)
-      case suffix         => (s"$base-$suffix",      s"-$suffix",       s"$base.v$date-$suffix-$sha", true)
+      // TODO: should this remain `-SNAPSHOT`? does `-SNAPSHOT` have a special meaning for maven / sbt? does `bin-SNAPSHOT` affect this?
+      // TODO: include `-bin` / `-pre` in the canonical version? in the osgi version?
```

Comment:
It's hard to tell, because there are so many systems that try to infer meaning from versions. Specifically: Maven, Ivy, the parts of Ivy that sbt injects, and sbt itself. And then there's OSGI which I _really_ don't know.

@cunei for instance (IIRC) was particularly concerned with using non-alphabetic sections in versions (I forgot the details).

In absence of some other, better informed, authority I would suggest experimentation and testing. :-/

With regards to strictly sbt's logic, I believe 2.12.2-bin-SNAPSHOT should work fine. The `-SNAPSHOT` would declare it as a non-stable version, and the `-bin` part would make it be considered binary compatible with the rest of the 2.12 series:

https://github.com/sbt/sbt/blob/v0.13.13/util/cross/src/main/input_sources/CrossVersionUtil.scala#L42